port_drayage_plugin/src/port_drayage_plugin.cpp

```diff
@@ -66,5 +77,99 @@ namespace port_drayage_plugin
 
         return 0;
     }
+
+    bool PortDrayagePlugin::plan_maneuver_cb(cav_srvs::PlanManeuversRequest &req, cav_srvs::PlanManeuversResponse &resp){
+        
+        geometry_msgs::PoseStamped stop_loc;
+
+        lanelet::BasicPoint2d current_loc(curr_pose_->pose.position.x, curr_pose_->pose.position.y);
+        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc, 1);
+
+        if(current_lanelets.size() == 0) {
+            ROS_WARN_STREAM("Cannot find any lanelet in map!");
+            return false;
+        }
+
+        auto current_lanelet = current_lanelets[0];
+        // uncomment after StopRule add.
+
+        // auto stop_rules = current_lanelet.second.regulatoryElementsAs<StopRule>();
+        // if(stop_rules.empty()){
+        //     return NULL;
+        // } else{
+        //     StopRule::Ptr stopRegelem = stop_rules.front();
+        //     // TODO get stopRegelem position and return the object with downtrack
+        //     stop_loc = geometry_msgs::PoseStampedConstPtr pose;
+        // }
+
+        // TODO
+        double stop_loc_downtrack = 0;
+        double current_loc_downtrack = wm_->routeTrackPos(current_loc).downtrack;
+
+        double speed_progress = _cur_speed->twist.linear.x;
+
+        double current_progress = current_loc_downtrack;
+
+        while(current_loc_downtrack < stop_loc_downtrack)
+        {
+            double end_dist = stop_loc_downtrack;
+            double dist_diff = end_dist - current_progress;
+
+            ros::Time start_time;
+            switch(req.prior_plan.maneuvers.back().type) {
+                case cav_msgs::Maneuver::LANE_FOLLOWING : start_time = req.prior_plan.maneuvers.back().lane_following_maneuver.end_time;
+                    break;
+                case cav_msgs::Maneuver::LANE_CHANGE : start_time = req.prior_plan.maneuvers.back().lane_change_maneuver.end_time;
+                    break;
+                case cav_msgs::Maneuver::INTERSECTION_TRANSIT_STRAIGHT : start_time = req.prior_plan.maneuvers.back().intersection_transit_straight_maneuver.end_time;
+                    break;
+                case cav_msgs::Maneuver::INTERSECTION_TRANSIT_LEFT_TURN : start_time = req.prior_plan.maneuvers.back().intersection_transit_left_turn_maneuver.end_time;
+                    break;
+                case cav_msgs::Maneuver::INTERSECTION_TRANSIT_RIGHT_TURN : start_time = req.prior_plan.maneuvers.back().intersection_transit_right_turn_maneuver.end_time;
+                    break;
+                case cav_msgs::Maneuver::STOP_AND_WAIT : start_time = req.prior_plan.maneuvers.back().stop_and_wait_maneuver.end_time;
+                    break;
+            }
+
+            resp.new_plan.maneuvers.push_back(
```

Comment:
If - for example - we have a long lanelet and a stop at the end of it I think we'd want to add a lane-keeping maneuver prior to the stop, if that makes sense. Should be able to calculate how long it'll take to stop, then put in a lane keeping maneuver for the distance prior.